### PR TITLE
Decode size-check hardening + graceful async runtime shutdown

### PR DESCRIPTION
**Pull request by tlmquintino**
Two production-hardening features for the Rust core / C++ async FFI, plus
test-coverage and documentation cleanup. All work is in the Rust core and
its C FFI; every language binding inherits the decode hardening for free.

### Structural descriptor ↔ payload size check on decode
The decode pipeline now rejects a data object whose descriptor claims a
decoded size that is structurally inconsistent with the frame's on-disk
payload, **before** any decompression or allocation runs.

- **Exact equality** (both directions) for the pipelines whose decoded
  size is fully determined by the descriptor:
  - uncompressed `encoding=none`: `num_values × dtype_width`
    (`ceil(num_values / 8)` for `bitmask`)
  - uncompressed `encoding=simple_packing`:
    `ceil(num_values × bits_per_value / 8)`
- Surfaces as a new `PipelineError::DescriptorSizeMismatch` naming the
  claimed vs. payload byte counts — a precise, operator-legible error
  instead of a downstream "failed to reserve".
- Also closes the gap where `simple_packing` silently tolerated extra
  trailing payload bytes.
- **Compressed codecs are deliberately not gated by a ratio heuristic.**
  Their decoded size cannot be derived without decompressing, and a fixed
  expansion-ratio ceiling would risk false-rejecting legitimately
  highly-compressible scientific data (constant fields, all-zero masks).
  They continue to fail gracefully at the fallible decompression
  allocation. Rationale recorded in `plans/DESIGN.md`.
- `estimate_decompressed_size` and the new gate share a single
  `exact_decoded_bytes()` source of truth (no drift).

### Graceful async runtime shutdown (`tgm_runtime_shutdown_blocking`)
Previously a no-op stub; now a real graceful shutdown.

- Blocks for up to `timeout_ms` while in-flight tasks drain, then returns
  the count of tasks that had not finished by the deadline (`0` on a clean
  drain) so callers can log/abort.
- **Single-shot:** once shut down the runtime is never rebuilt, every
  subsequent `tgm_async_*` call fails fast with `TGM_ERROR_IO`, and a
  second shutdown is an idempotent no-op.
- Internals: the runtime singleton moved from `OnceLock` to a takeable
  `RuntimeState` machine (so `shutdown_timeout` can consume the owned
  runtime); a `LIVE_TASKS` counter with a drop-guard at the spawn choke
  point tracks not-yet-drained work; the count is captured before forced
  teardown so it stays accurate.
- **C ABI unchanged** — `tensogram.h` regenerated by cbindgen.

## Added (tests)
- `RleCompressor` / `RoaringCompressor` wrapper unit tests
  (round-trip, all-zero/all-set, short-blob and size-mismatch errors,
  range-unsupported): **0% → ~94%** line coverage on both files.
- `MaskMethod` `name()`/`from_name()` registry round-trip, default,
  unknown-method rejection, feature-gated blosc2, and `MaskError` Display
  strings: `bitmask/mod.rs` **44% → 97%**.
- Empty-mask round-trips across the mask codecs (none/lz4/zstd/blosc2).
- Rust unit test + isolated C++ GoogleTest
  (`cpp/tests/test_async_shutdown_during_flight.cpp`) for the shutdown
  drain/count and post-shutdown rejection.

## Fixed
- `tgm_runtime_shutdown_blocking` no longer ignores `timeout_ms` and no
  longer always reports zero unfinished tasks.

## Docs / cleanup
- Retired `PLAN_FORTRAN.md` (the synchronous Fortran binding has shipped)
  and scrubbed all 30 references across docs, Fortran source/tests, CMake,
  CI, and the plans.
- Documented the size-mismatch edge case (`docs/src/edge-cases.md`) and
  the real shutdown contract (`docs/src/guide/cpp-async.md`); removed the
  two completed `plans/TODO.md` entries; removed a stray local `.coverage`
  artifact.

## Verification
- `cargo fmt --check`, `cargo clippy --workspace --all-targets --all-features -D warnings`
- `cargo test --workspace`, `cargo test -p tensogram --features remote,async`
- `ruff check`/`format --check`; pytest: core (597), xarray (245), zarr (235)
- C++ `ctest`: 182/182 (incl. the new shutdown test)
- `mdbook build docs/`

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-128

### CODE REVIEW

File: rust/tensogram-ffi/src/async_core.rs [hunk 1]
```diff
@@ -331,7 +405,15 @@ where
...
 
+    // Count this task as live before it is spawned.  The drop-guard
+    // inside the future decrements on every exit path (completion,
+    // inner error, cancel, timeout, or the future being dropped when
+    // the runtime is torn down), so `LIVE_TASKS` is an accurate count
+    // of not-yet-drained tasks for `tgm_runtime_shutdown_blocking`.
+    LIVE_TASKS.fetch_add(1, Ordering::SeqCst);
+
     rt.spawn(async move {
+        let _live = LiveTaskGuard;
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/src/async_core.rs]: Fixed in 7116c29. The guard is now built via `LiveTaskGuard::new()` *before* `rt.spawn` and moved into the future, so its `Drop` runs even when the task is dropped before its first poll (the runtime-torn-down-while-queued case) — keeping `LIVE_TASKS` accurate.

File: rust/tensogram-ffi/src/async_core.rs [hunk 2]
```diff
@@ -1231,23 +1313,78 @@ pub extern "C" fn tgm_runtime_configure(
...
+
+    // Cooperative drain.  Poll `LIVE_TASKS` until it reaches zero or
+    // the deadline elapses, then capture the residual *before* the
+    // forced teardown below.  Reading the count here — rather than
+    // after `shutdown_timeout` — is what makes the result accurate:
+    // dropping the runtime drops every still-pending future, which
+    // runs each `LiveTaskGuard` and would otherwise drive the count
+    // to zero regardless of whether those tasks actually finished.
+    let deadline = std::time::Instant::now() + Duration::from_millis(timeout_ms);
+    let poll_interval = Duration::from_millis(5);
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/src/async_core.rs]: Fixed in 7116c29. The drain loop no longer does `Instant + Duration`; it compares `start.elapsed()` against the timeout, which is panic-free for any `u64` (the remaining-budget subtraction is guarded by the `elapsed >= timeout` break).

File: rust/tensogram-ffi/tensogram.h [hunk 3]
```diff
@@ -1243,19 +1243,30 @@ tgm_error tgm_runtime_configure(uint32_t workers,
...
- * accepted but ignored today.
+ * Shut the shared async runtime down, blocking for up to
+ * `timeout_ms` while in-flight tasks drain.
+ *
+ * Returns the number of tasks that had **not** finished when the
+ * timeout elapsed (`0` on a clean drain).  After this call the
+ * runtime is permanently shut down: every subsequent `tgm_async_*`
+ * entry point fails fast with [`TgmError::Io`] and a descriptive
+ * [`tgm_last_error`](crate::tgm_last_error) (`"runtime has been shut
+ * down"`).  The runtime is single-shot — there is no rebuild.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/tensogram.h]: Fixed in 030cd93. The source doc comment now uses plain C references (`TGM_ERROR_IO`, `tgm_last_error()`) instead of rustdoc intra-doc links; the header was regenerated by cbindgen.

File: cpp/tests/test_async_shutdown_during_flight.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    for (std::size_t i = 0; i < payload.size(); ++i) {
+        payload[i] = static_cast<std::uint8_t>(i);
+    }
+    std::vector<std::pair<const std::uint8_t*, std::size_t>> objs;
+    objs.emplace_back(payload.data(), payload.size());
+
+    auto bytes = tensogram::encode(meta_json, objs);
+    auto* fp = std::fopen(path.c_str(), "wb");
+    std::fwrite(bytes.data(), 1, bytes.size(), fp);
+    std::fclose(fp);
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply cpp/tests/test_async_shutdown_during_flight.cpp:66]: Fixed in 38a6ea7. `make_sync_test_file()` now null-checks the `fopen` result and verifies the `fwrite` size, throwing `std::runtime_error` (which GoogleTest reports as a clear failure) instead of segfaulting.

File: rust/tensogram-ffi/src/async_core.rs [hunk 5]
```diff
@@ -1342,3 +1508,64 @@ fn dtype_name(d: tensogram::dtype::Dtype) -> &'static str {
...
+            tgm_runtime_shutdown_blocking(10),
+            0,
+            "second shutdown must be a no-op returning 0"
+        );
+
+        // The leaked task handle is intentionally not freed/joined: the
+        // runtime that owned its future is gone, so joining would block
+        // forever.  Leaking one Arc in a unit test is acceptable.
+        let _ = task;
+    }
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/src/async_core.rs]: Fixed in f9ddb92. Confirmed `tgm_async_task_free` only drops the handle box (one `Arc<TaskShared>` ref) and never joins/blocks, so the test now frees the handle instead of leaking it. The test still passes without hanging.

File: rust/tensogram-ffi/src/async_core.rs [hunk 6]
```diff
@@ -331,7 +416,23 @@ where
...
+    // async block would skip the decrement in that drop-before-poll
+    // case, permanently inflating the counter.  `LiveTaskGuard::new`
+    // does the `fetch_add`; its `Drop` does the matching `fetch_sub`.
+    let live_guard = LiveTaskGuard::new();
+
     rt.spawn(async move {
+        // Bind the guard into the future so its lifetime is the
+        // future's lifetime.  `let _live = live_guard;` rather than a
+        // bare move expression so the binding is unambiguous.
+        let _live = live_guard;
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/src/async_core.rs:471]: Fixed in 78e18e4. Added a `CompletionGuard` RAII backstop (built before `spawn`, moved into the future like `LiveTaskGuard`) that calls `complete(AsyncOutcome::Cancelled)` on drop. Since `complete()` is idempotent it is a no-op on the happy path and only fires when the future is dropped without completing, so the task always leaves `Pending` and joiners wake. The shutdown unit test now joins the stranded task and asserts `Cancelled` (it would hang, not just fail, if this regressed).

File: rust/tensogram-ffi/tests/async_core_tests.rs [hunk 7]
```diff
@@ -954,12 +954,10 @@ fn async_decode_range_round_trip() {
...
-#[test]
-fn async_runtime_shutdown_blocking_is_noop() {
-    // Reserved ABI: currently a no-op that reports zero unfinished tasks
-    // (process exit is abrupt by design).  Must not panic.
-    assert_eq!(tgm_runtime_shutdown_blocking(100), 0);
-}
+// NOTE: `tgm_runtime_shutdown_blocking` tears down the process-global
+// runtime permanently, so it cannot be exercised here without poisoning
+// every sibling test in this binary.  Its tests live in the dedicated,
+// process-isolated `tests/async_shutdown_tests.rs` binary.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-ffi/tests/async_core_tests.rs]: Fixed in b5ed70d. The note now points to the actual coverage: the in-crate unit test `async_core::tests::shutdown_blocking_drains_and_reports_unfinished` and the C++ gtest, not a non-existent binary.

File: rust/tensogram-encodings/src/pipeline.rs [hunk 8]
```diff
@@ -1712,4 +1819,250 @@ mod tests {
...
+                swap_unit_size: 1,
+                compression_backend: CompressionBackend::default(),
+                intra_codec_threads: 0,
+                compute_hash: false,
+            };
+            let payload = encode_pipeline(&data, &cfg).unwrap().encoded_bytes;
+            // Compressed payload length differs from the decoded length;
+            // the gate must not fire.
+            assert_ne!(payload.len(), data.len());
+            let out = decode_pipeline(&payload, &cfg, false).expect("compressed must decode");
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram-encodings/src/pipeline.rs]: Fixed in b92c127. Replaced the codec-dependent `assert_ne!` on a small mixed input with a 64 KiB all-zero buffer that provably compresses below its raw length on any zstd build, asserting `payload.len() < data.len()`. The test still verifies the real property (the exact-size gate is skipped for compressed pipelines and the object round-trips).
